### PR TITLE
[CSL-2227] Add Support for Sort Configurations

### PR DESCRIPTION
**Pull request by Mudaafi**

### CODE REVIEW

File: src/constructor.io/client/modules/Catalog.cs
Comment (jjl014) [inline src/constructor.io/client/modules/Catalog.cs:1062]: Similar to my comment above, it seems a bit complex to need to keep track of all these request classes. Should we do something similar to how our facet functions work?

In this case `SetSortOptions` would accept two parameters, a list of sort options and a section (that defaults to 'Products')

File: src/constructor.io/client/modules/Catalog.cs
Comment (jjl014) [inline src/constructor.io/client/modules/Catalog.cs:1091]: Should this be `DeleteSortOptions`(plural)?

And similar to the previous comments, should we just simplify this to just accept two arguments? (list of sort options to remove and section)

**Review by jjl014 — COMMENTED** [review]
I think it might be worth having another discussion on how we want to implement this. I feel like having all of the different request classes might be unnecessary here.

I left a couple of comments, but holding off on a full review of the PR until after we talk through the changes here. Would love to get your thoughts on the comments when your back!

File: src/constructor.io/client/modules/Catalog.cs
Comment (Mudaafi) [reply src/constructor.io/client/modules/Catalog.cs:1032]: I've removed this class and merged it with the base class per our discussion.

File: src/constructor.io/client/modules/Catalog.cs
Comment (Mudaafi) [reply src/constructor.io/client/modules/Catalog.cs:1062]: Following an in-depth discussion, we're running with the Request Object pattern. See [Slack Thread Here](https://constructor.slack.com/archives/C02RSEE9QE4/p1683741624059059).

File: src/constructor.io/client/modules/Catalog.cs
Comment (Mudaafi) [reply src/constructor.io/client/modules/Catalog.cs:1091]: Ooh nice catch on the missing `s`. As for the simplification, we're running with the Request Object pattern. See [Slack Thread Here](https://constructor.slack.com/archives/C02RSEE9QE4/p1683741624059059).

File: src/constructor.io/models/Catalog/SortOptionsSingleRequest.cs
Comment (jjl014) [inline src/constructor.io/models/Catalog/SortOptionsSingleRequest.cs:6]: Should we add `for a single Sort Option` at the end of this description?

File: src/constructor.io/models/common/SortOptionForModifyRequest.cs
Comment (jjl014) [inline src/constructor.io/models/common/SortOptionForModifyRequest.cs:13]: Typo `/` => `.`

**Review by jjl014 — COMMENTED** [review]
Great job and thanks for working on the updates/refactoring! I left some comments on a couple things.

I'm still not the biggest fan of the three different request classes, but I can't think of a better way to represent them especially since they're being used across different endpoints/methods (i.e. GET, PATCH, etc). I think you did an awesome job handling the different cases!

Would love to also get some additional feedback from the rest of the @Constructor-io/customer-integrations team.

File: src/constructor.io/models/common/SortOptionList.cs
Comment (Mudaafi) [reply src/constructor.io/models/common/SortOptionList.cs:11]: Ah, this class exists for two reasons, both related to request handling:
1. In `RetrieveSortOptions` and `SetSortOptions`, this class represents the response structure that is returned from the API so it has to exist unless we map the response to a generic `Hashtable` and extract the `List<SortOption>` manually before returning it, not a pattern we've used in this repo.

2. In `SetSortOptions`, this class is also the request object required by the API. An alternative is using a `Hashtable` to map the `List<SortOption>` before we send it along the request, similar to how we do it in `PatchSearchabilities` but it seemed like extra work when we can just define a class. 

Open to change it though, lmk your thoughts

File: src/constructor.io/models/common/SortOptionForCreateOrReplace.cs
Comment (Mudaafi) [reply src/constructor.io/models/common/SortOptionForCreateOrReplace.cs]: mb, fixed!

File: src/constructor.io/models/common/SortOption.cs
Comment (Mudaafi) [reply src/constructor.io/models/common/SortOption.cs:31]: Hmm you're right, that makes more sense and keeps it aligned with the API. I'll make the changes.

File: src/constructor.io/models/common/SortOptionForModifyRequest.cs
Comment (Mudaafi) [reply src/constructor.io/models/common/SortOptionForModifyRequest.cs:13]: Nice catch

File: src/constructor.io/models/common/SortOptionForModifyRequest.cs
Comment (Mudaafi) [reply src/constructor.io/models/common/SortOptionForModifyRequest.cs:9]: You are absolutely right. I totally forgot that was a thing. I will have to mark the `SortOrder` attribute to be optional on the `SortOption` class though but I'm guessing that should be fine now since we're using the same class to send delta updates.

File: src/constructor.io/models/common/SortOptionList.cs
Comment (jjl014) [reply src/constructor.io/models/common/SortOptionList.cs:11]: Ooo, that makes sense! I forgot that we also need response models in addition to the requests. 🤦‍♂️ 

Sorry for the dumb comment. Thank you for the detailed explanation!

File: src/constructor.io/models/common/SortOptionForModifyRequest.cs
Comment (jjl014) [reply src/constructor.io/models/common/SortOptionForModifyRequest.cs]: All good, if it seems weird I think we can just keep this class since it's only used internally I believe.

File: src/constructor.io/models/common/SortOptionList.cs
Comment (Mudaafi) [reply src/constructor.io/models/common/SortOptionList.cs:11]: No worries! I appreciate the detailed review & discussion, wouldn't want to merge redundant code 😁

File: src/constructor.io/models/common/SortOptionForModifyRequest.cs
Comment (Mudaafi) [reply src/constructor.io/models/common/SortOptionForModifyRequest.cs]: No worries, I think having a single `SortOption` class makes things more readable + developer friendly anyways since it's more intuitive to expect the same `SortOption` class to be used but for null values to be ignored. I made the changes but if anyone has a differing opinion, feel free to voice out.

The contrasting examples to this pattern are the different request objects we have, but those act as sort of "recipes" that inform the SDK consumer what can/cannot be passed to a specific function/api call so it makes more sense to have different classes that only contain the attributes specific to those calls.

**Review by jjl014 — APPROVED** [review]
Sweet, thanks for making the updates. This is looking great to me!

@Constructor-io/customer-integrations Would love to get another pair of eyes in case I missed anything or if anyone has any additional thoughts. Thanks!

Otherwise, I think this is ready to go. 👍

**Review by crgee1 — APPROVED** [review]
lgtm

**Pull request merged**
